File: android/src/main/java/com/lightningkite/khrysalis/rx/RxExtensions.actual.kt
package com.lightningkite.khrysalis.rx

import com.lightningkite.khrysalis.Box
import com.lightningkite.khrysalis.SwiftName
import com.lightningkite.khrysalis.observables.MutableObservableProperty
import com.lightningkite.khrysalis.observables.ObservableProperty
import com.lightningkite.khrysalis.post
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.functions.BiFunction


/**
 * Combine latest operator
 */
fun <Element : Any, R : Any, OUT : Any> Observable<Element>.combineLatest(
    observable: Observable<R>,
    function: (Element, R) -> OUT
): Observable<OUT> = Observable.combineLatest(this, observable, BiFunction(function))

fun <IN : Any, OUT : Any> List<Observable<IN>>.combineLatest(
    combine: (List<IN>) -> OUT
): Observable<OUT> = Observable.combineLatest(this) { stupidArray: Array<Any?> ->
    combine(stupidArray.toList() as List<IN>)
}
fun <IN : Any> List<Observable<IN>>.combineLatest(): Observable<List<IN>>
        = Observable.combineLatest(this) { stupidArray: Array<Any?> -> stupidArray.toList() as List<IN>}

fun <Element> Observable<Box<Element>>.filterNotNull(): Observable<Element> =
    this.filter { it.value != null }.map { it.value }

fun <Element : Any, Destination : Any> Observable<Element>.mapNotNull(transform: (Element) -> Destination?): Observable<Destination> =
    this.flatMap {
        val result = transform(it)
        if (result == null)
            Observable.empty()
        else
            Observable.just(result)
    }

fun <Element : Any> Single<Element>.working(observable: MutableObservableProperty<Boolean>): Single<Element> {
    return this
        .doOnSubscribe { it -> post { observable.value = true } }
        .doFinally { post { observable.value = false } }
}

File: ios/Khrysalis/src/rx/RxExtensions.actual.swift
import RxSwift
import RxRelay

//--- Observable<Element>.combineLatest(Observable<R>, (Element,R)->OUT)
extension ObservableType {
    /**
    Merges the specified observable sequences into one observable sequence by using the selector function whenever any of the observable sequences produces an element.
    - seealso: [combineLatest operator on reactivex.io](http://reactivex.io/documentation/operators/combinelatest.html)
    - parameter resultSelector: Function to invoke whenever any of the sources produces an element.
    - returns: An observable sequence containing the result of combining elements of the sources using the specified result selector function.
    */
    public func combineLatest<O2: ObservableType, OUT>
        (observable: O2, function: @escaping (Element, O2.Element) throws -> OUT)
            -> Observable<OUT> {
                return Observable<OUT>.combineLatest(self, observable, resultSelector: function)
    }
    public func combineLatest<O2: ObservableType, OUT>
        (_ observable: O2, _ function: @escaping (Element, O2.Element) throws -> OUT)
            -> Observable<OUT> {
                return Observable<OUT>.combineLatest(self, observable, resultSelector: function)
    }
}

//--- Observable<Element>.filterNotNull()
extension Observable where Observable.Element: OptionalConvertible {
    func filterNotNull() -> Observable<Element.Wrapped> {
        self.filter { $0.asOptional != nil }.map { $0.asOptional! }
    }
}

//--- List<Observable<IN>>.combineLatest((List<IN>)->OUT)
func xListCombineLatest<IN, OUT>(
    _ self: Array<Observable<IN>>,
    combine: @escaping (Array<IN>) -> OUT
) -> Observable<OUT> {
    return Observable.combineLatest(self, resultSelector: combine)
}
func xListCombineLatest<IN>(
    _ self: Array<Observable<IN>>
) -> Observable<Array<IN>> {
    return Observable.combineLatest(self)
}
extension Array where Element: ObservableType {
    func combineLatest<OUT>(combine: @escaping (Array<Element.Element>)->OUT) -> Observable<OUT> {
        return Observable.combineLatest(self, resultSelector: combine)
    }
    func combineLatest() -> Observable<Array<Element.Element>> {
        return Observable.combineLatest(self)
    }
}

//--- Observable<Element>.mapNotNull((Element)->Destination?)
public extension Observable {
    func mapNotNull<Destination>(_ transform: @escaping (Element) -> Destination?) -> Observable<Destination> {
        return mapNotNull(transform: transform)
    }
    func mapNotNull<Destination>(transform: @escaping (Element) -> Destination?) -> Observable<Destination> {
        return self.flatMap { (it: Element) -> Observable<Destination> in
            if let result: Destination = transform(it) {
                return Observable<Destination>.just(result)
            } else {
                return Observable<Destination>.empty()
            }
        }
    }
}

//--- Single<Element>.working(MutableObservableProperty<Boolean>)
public extension PrimitiveSequenceType where Trait == SingleTrait {
    func working(_ observable: MutableObservableProperty<Bool>) -> Single<Element> {
        return self.do(onSuccess: { _ in observable.value = false }, onError: { _ in observable.value = false }, onSubscribe: { observable.value = true })
    }
    func working(observable: MutableObservableProperty<Bool>) -> Single<Element> {
        return working(observable)
    }
}
